Public/Src/Engine/Processes/SharedOpaqueOutputHelper.cs

```diff
@@ -10,31 +11,140 @@
+                var writeAttributesDenied = !FileUtilities.HasWritableAttributeAccessControl(expandedPath);
+                if (writeAttributesDenied)
+                {
+                    FileUtilities.SetFileAccessControl(expandedPath, FileSystemRights.WriteAttributes | FileSystemRights.WriteExtendedAttributes, allow: true);
+                }
+
+                try
+                {
+                    // Only the creation time is used to identify a file as the output of a shared opaque
+                    FileUtilities.SetFileTimestamps(expandedPath, new FileTimestamps(WellKnownTimestamps.OutputInSharedOpaqueTimestamp));
+                }
+                catch (BuildXLException ex)
+                {
+                    // On (unsafe) double writes, a race can occur, we give the other contender a second chance
+                    if (IsSharedOpaqueOutput(expandedPath))
+                    {
+                        return;
+                    }
+
+                    // Since these files should be just created outputs, this shouldn't happen and we bail out hard.
+                    throw new BuildXLException(I($"Failed to open output file '{expandedPath}' for writing."), ex);
+                }
+                finally
+                {
+                    // Restore the attributes as they were originally set
+                    if (writeAttributesDenied)
+                    {
+                        FileUtilities.SetFileAccessControl(expandedPath, FileSystemRights.WriteAttributes | FileSystemRights.WriteExtendedAttributes, allow: false);
+                    }
+                }
+            }
+
+            /// <summary>
+            /// Checks if the given path is an output under a shared opaque by verifying whether <see cref="WellKnownTimestamps.OutputInSharedOpaqueTimestamp"/> is the creation time of the file
+            /// </summary>
+            /// <remarks>
+            /// If the given path is a directory, it is always considered part of a shared opaque
+            /// </remarks>
+            public static bool IsSharedOpaqueOutput(string expandedPath)
             {
-                // Only the creation time is used to identify a file as the output of a shared opaque
-                FileUtilities.SetFileTimestamps(expandedPath, new FileTimestamps(WellKnownTimestamps.OutputInSharedOpaqueTimestamp));
+                try
+                {
+                    var creationTime = FileUtilities.GetFileTimestamps(expandedPath).CreationTime;
+                    return creationTime == WellKnownTimestamps.OutputInSharedOpaqueTimestamp;
+                }
+                catch (BuildXLException ex)
+                {
+                    throw new BuildXLException(I($"Failed to open output file '{expandedPath}' for reading."), ex);
+                }
             }
-            catch (BuildXLException ex)
+        }
+
+        private static unsafe class Unix

```

Comment:
>Unix [](start = 36, length = 4)

If we are doing this change, why not doing it for both Win and Unix? It will make the code simpler, and considering this is a breaking change, better do it now that later.

Reply:
That's exactly why: I don't really want this breaking change, I only want to validate if this fixes the problem for Mac builds.  And indeed, there doesn't seem to be anything wrong with using timestamps on Windows (because of timestamp faking) so we don't need to make this a breaking change on Windows.

Reply:
That's fine, and I won't push hard for it. But I'd rather have a uniform marker. Less cognitive load and simpler code.
I don't expect this breaking change to be a problematic one: there are no real MSBuild customers yet, and WDG always scrubs everything themselves before building. So should be almost a noop.

---
In reply to: [332139469](https://github.com/microsoft/BuildXL/pull/1007#discussion_r332139469) [](ancestors = 332139469)

Reply:
Thanks for all those pointers.

I'll leave a comment saying that we should unify this eventually.